# Description

So, we work in oil company and need to decide where to drill the new well

We have oil samples in three regions(10,000 fields in one). 

Build a machine learning model to help determine the region where mining will get the most profit. Analyze the potential rewards and risks with *Bootstrap*.

# 1. About data

In [1]:
#library import
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import roc_curve, mean_squared_error,mean_absolute_error, accuracy_score,r2_score, confusion_matrix, recall_score,roc_auc_score, precision_score, f1_score

In [2]:
#check data zero
geo_data_0 = pd.read_csv('geo_data_0.csv')
display(geo_data_0.head())

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [3]:
geo_data_0.info()
geo_data_0.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


0

In [4]:
##check data one
geo_data_1 = pd.read_csv('geo_data_1.csv')
display(geo_data_1.head())

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [5]:
geo_data_1.info()
geo_data_1.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


0

In [6]:
##check data two
geo_data_2 = pd.read_csv('geo_data_2.csv')
display(geo_data_2.head())

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [7]:
geo_data_2.info()
geo_data_2.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


0

### Сonclusion

There are 3 files with geological prospecting data  from three regions. 
There are no duplicates or null values.

100000 entries and 5 columns for all:

id — inique well id 

f0, f1, f2 — doesn't matter what it mean, but they are important

product — well reserves (thousand barrels).

Important:
- Linear regression only 
- We check 500 wells and choose 200 best.
- Budget for region — 10 000 000 000 roubles.
- 1 barrel = 450 roubles income.

We need to find regions, where probability of losses will be less than 2.5%. After we should choose region with the highest average profit.

# 2. Research

##### Region 0

In [8]:
#features&target
features_0 = geo_data_0.drop(['product','id'], axis=1)
target_0 = geo_data_0['product']

#numeric
numeric = ['f0','f1','f2']

scaler = StandardScaler()
scaler.fit(features_0[numeric])
features_0[numeric] = scaler.transform(features_0[numeric])
display(features_0.shape)

(100000, 3)

In [9]:
#separate for two
features_train_0, features_valid_0, target_train_0, target_valid_0 = train_test_split(
    features_0, target_0, test_size=0.25, random_state=12345)

display(features_train_0.shape)
display(features_valid_0.shape)

(75000, 3)

(25000, 3)

In [10]:
#RMSE и mean stock of raw materials
model = LinearRegression()
model.fit(features_train_0, target_train_0)
predictions_valid_0 = model.predict(features_valid_0)

rmse_0 = (mean_squared_error(target_valid_0,predictions_valid_0))**0.5
print('RMSE:', rmse_0)

mean_product_0 = model.predict(features_valid_0).mean() 
print('stock of raw materials:', mean_product_0)

RMSE: 37.5794217150813
stock of raw materials: 92.59256778438038


##### Region 1

In [11]:
#features&target
features_1 = geo_data_1.drop(['product','id'], axis=1)
target_1 = geo_data_1['product']

#numeric
numeric = ['f0','f1','f2']

scaler = StandardScaler()
scaler.fit(features_1[numeric])
features_1[numeric] = scaler.transform(features_1[numeric])
display(features_1.shape)

(100000, 3)

In [12]:
#separate for two
features_train_1, features_valid_1, target_train_1, target_valid_1 = train_test_split(
    features_1, target_1, test_size=0.25, random_state=12345)

display(features_train_1.shape)
display(features_valid_1.shape)

(75000, 3)

(25000, 3)

In [13]:
#stock of raw materials
model = LinearRegression()
model.fit(features_train_1, target_train_1)
predictions_valid_1 = model.predict(features_valid_1)

rmse_1 = (mean_squared_error(target_valid_1,predictions_valid_1))**0.5
print('RMSE:', rmse_1)

mean_product_1 = model.predict(features_valid_1).mean() 
print('stock of raw materials:', mean_product_1)

RMSE: 0.893099286775616
stock of raw materials: 68.72854689544602


##### Region 2

In [14]:
#features&target
features_2 = geo_data_2.drop(['product', 'id'], axis=1)
target_2 = geo_data_2['product']

#numeric
numeric = ['f0','f1','f2']

scaler = StandardScaler()
scaler.fit(features_2[numeric])
features_1[numeric] = scaler.transform(features_2[numeric])
display(features_2.shape)

(100000, 3)

In [15]:
#separate for two

features_train_2, features_valid_2, target_train_2, target_valid_2 = train_test_split(
    features_2, target_2, test_size=0.25, random_state=12345)

display(features_train_2.shape)
display(features_valid_2.shape)

(75000, 3)

(25000, 3)

In [16]:
#stock of raw materials
model = LinearRegression()
model.fit(features_train_2, target_train_2)
predictions_valid_2 = model.predict(features_valid_2)

rmse_2 = (mean_squared_error(target_valid_2,predictions_valid_2))**0.5
print('RMSE:', rmse_2)

mean_product_2 = model.predict(features_valid_2).mean() 
print('stock of raw materials:', mean_product_2)

RMSE: 40.02970873393434
stock of raw materials: 94.96504596800489


In [17]:
#compare data
rmse_mean_geo = pd.DataFrame({'geo_data':['geo_data_0','geo_data_1','geo_data_2'], 'rmse':[rmse_0, rmse_1, rmse_2], 'mean product':[mean_product_0,mean_product_1,mean_product_2]})

display(rmse_mean_geo)

geo_data       rmse  mean product
0  geo_data_0  37.579422     92.592568
1  geo_data_1   0.893099     68.728547
2  geo_data_2  40.029709     94.965046

### Сonclusion

For three:
- delete id column
- all data started from 0
- train/valid and numeric for all
- rmse and stock of raw materials for all

Now regions geo_data_2 and geo_data_0 are best, because they have highest mean product. 
But region geo_data_1 has lowest rmse.
We need to think.

# 3. Prepare for profit

In [18]:
#for all 3 fields
budget_region = 10000000000 #budget for region
budget_slit = budget_region / 200 #budget for slit
income_barrel = 450 #income from barrel
income_one = 450000 #income from 1000 barrels
volume = budget_slit / income_one #the volume of raw materials for the break-even development of a new well

In [19]:
#mean of raw material in region 
mean_product_0 = geo_data_0['product'].mean()
print('mean of raw material in region  0:',mean_product_0)

mean_product_1 = geo_data_1['product'].mean()
print('mean of raw material in region  1:',mean_product_1)

mean_product_2 = geo_data_2['product'].mean()
print('mean of raw material in region  2:',mean_product_2)

print('volume of raw materials for the break-even development of a new well:', volume)

mean of raw material in region  0: 92.49999999999976
mean of raw material in region  1: 68.82500000002561
mean of raw material in region  2: 95.00000000000041
volume of raw materials for the break-even development of a new well: 111.11111111111111


### Сonclusion

The volume of raw materials for the break-even development of a new well is slightly higher than the average stock of raw materials.

But this is only an average stock, so we still have only 2 priorityregions: geo_data_0 and geo_data_2. 

Let's see what's next.

# 4. Function for profit and loss

#### Region 0

In [20]:
#column for predict
predict_prod_0 = pd.Series(model.predict(features_0))
geo_data_0['predict_product'] = predict_prod_0
display(geo_data_0.head())

id        f0        f1        f2     product  predict_product
0  txEyH  0.705745 -0.497823  1.221170  105.280062        78.593786
1  2acmU  1.334711 -0.340164  4.365080   73.037750        84.128220
2  409Wp  1.022732  0.151990  1.419926   85.265647        78.900387
3  iJLyR -0.032172  0.139033  2.978566  168.620776        81.603233
4  Xdl7t  1.988431  0.155413  4.751769  154.036647        84.789854

In [21]:
#function for profit
def profit(target_0, features_0):
    features_sorted = features_0.sort_values(ascending=False)
    selected_points = target_0[features_sorted.index][:200]
    product = selected_points.sum()
    prof = product * income_one
    cost = budget_slit * 200
    return prof - cost

In [22]:
#value select
state = np.random.RandomState(12345)   
tar_0  = target_valid_0.reset_index(drop=True)
pred_0 = pd.Series(predictions_valid_0)
profit_values_0 = []

for i in range(1000):
        target_sample = tar_0.sample(500, replace=True, random_state=state)
        features_sample = pred_0[target_sample.index]
        profit_values_0.append(profit(target_sample, features_sample))
profit_values_0 = pd.Series(profit_values_0)

mean_0 = profit_values_0.mean()

#confidence interval
confidence_interval_0 = (profit_values_0.quantile(0.025), profit_values_0.quantile(0.975))

#risk
risk_0= ((profit_values_0 < 0).mean()* 100)

print("profit:", mean_0)
print("95% confidence interval:", confidence_interval_0)
print("risk in region 0: ", round(risk_0,2),"%")

profit: 425938526.910592
95% confidence interval: (-102090094.83793654, 947976353.358369)
risk in region 0:  6.0 %


#### Region 1

In [23]:
#column for predict
predict_prod_1 = pd.Series(model.predict(features_1))
geo_data_1['predict_product'] = predict_prod_1
display(geo_data_1.head())

id         f0         f1        f2     product  predict_product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103        75.269514
1  62mP7  14.272088  -3.475083  0.999183   26.953261        72.515536
2  vyE1P   6.263187  -5.948386  5.001160  134.766305        67.490944
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408        78.257555
4  AHL4O  12.702195  -8.147433  5.004363  134.766305        86.319923

In [24]:
#function for profit
def profit(target_1, features_1):
    features_sorted = features_1.sort_values(ascending=False)
    selected_points = target_1[features_sorted.index][:200]
    product = selected_points.sum()
    prof = product * income_one
    cost = budget_slit * 200
    return prof - cost

#value select
state = np.random.RandomState(12345)   
tar_1  = target_valid_1.reset_index(drop=True)
pred_1 = pd.Series(predictions_valid_1)
profit_values_1 = []

for i in range(1000):
        target_sample = tar_1.sample(500, replace=True, random_state=state)
        features_sample = pred_1[target_sample.index]
        profit_values_1.append(profit(target_sample, features_sample))
profit_values_1 = pd.Series(profit_values_1)
mean_1 = profit_values_1.mean()

print("profit:", mean_1)

confidence_interval_1 = (profit_values_1.quantile(0.025), profit_values_1.quantile(0.975))
risk_1= ((profit_values_1 < 0).mean()* 100)

print("95% confidence interval:", confidence_interval_1)
print("risk in region 1: ", round(risk_1,2),"%")

profit: 515222773.44329005
95% confidence interval: (68873225.37050177, 931547591.2570496)
risk in region 1:  1.0 %


#### Region 2

In [25]:
#column for predict
predict_prod_2 = pd.Series(model.predict(features_2))
geo_data_2['predict_product'] = predict_prod_2
display(geo_data_2.head())

id        f0        f1        f2     product  predict_product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673        75.968506
1  WJtFt  0.262778  0.269839 -2.530187   56.069697        66.329664
2  ovLUW  0.194587  0.289035 -5.586433   62.871910        48.880039
3  q6cA6  2.236060 -0.553760  0.930038  114.572842        86.178035
4  WPMUX -0.515993  1.716266  5.899011  149.600746       114.363434

In [26]:
#function for profit
def profit(target_2, features_2):
    features_sorted = features_2.sort_values(ascending=False)
    selected_points = target_2[features_sorted.index][:200]
    product = selected_points.sum()
    prof = product * income_one
    cost = budget_slit * 200
    return prof - cost


state = np.random.RandomState(12345)   
tar_2  = target_valid_2.reset_index(drop=True)
pred_2 = pd.Series(predictions_valid_2)
profit_values_2 = []

for i in range(1000):
        target_sample = tar_2.sample(500, replace=True, random_state=state)
        features_sample = pred_2[target_sample.index]
        profit_values_2.append(profit(target_sample, features_sample))
profit_values_2 = pd.Series(profit_values_2)
mean_2 = profit_values_2.mean()

print("profit:", mean_2)

confidence_interval_2 = (profit_values_2.quantile(0.025), profit_values_2.quantile(0.975))
risk_2= ((profit_values_2 < 0).mean()* 100)

print("95% confidence interval:", confidence_interval_2)
print("risk in region 2: ", round(risk_2,2),"%")

profit: 435008362.78275585
95% confidence interval: (-128880547.32978901, 969706954.1802679)
risk in region 2:  6.4 %


# Final conclusion

So, here we should choose region 1.
Yes, it doesn't win in previous calculations, but it has only 1% risk!

Yes, profit(515222773) is not so big, but we think about future.